File: backend/app/schemas/calendar.py
from pydantic import BaseModel, Field
from datetime import date, time
from typing import Optional
from app.schemas.subject import SubjectResponse

class CalendarEventBase(BaseModel):
    date: date
    event_type: Optional[str] = Field(default="holiday", description="'holiday', 'working_day_override', 'college_closure', 'exam_break', 'working_saturday', 'exam', 'exam_day'")
    description: Optional[str] = None
    timetable_day_override: Optional[int] = None
    subject_id: Optional[int] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    title: Optional[str] = None
    category: Optional[str] = None
    schedule_effect: Optional[str] = None
    end_date: Optional[date] = None

class CalendarEventCreate(CalendarEventBase):
    pass

class CalendarEventResponse(CalendarEventBase):
    id: int
    semester_id: int
    subject: Optional[SubjectResponse] = None

    class Config:
        from_attributes = True
        json_encoders = {
            time: lambda t: t.strftime("%H:%M:%S")
        }

File: frontend/src/services/calendar.ts
import api from "./api";
import type { Subject } from "./subject";

export interface CalendarEvent {
  id: number;
  semester_id: number;
  date: string; // "YYYY-MM-DD"
  event_type: string; // "holiday", "working_day_override", "college_closure", "exam_break", "working_saturday", "exam", "exam_day"
  description?: string;
  timetable_day_override?: number;
  subject_id?: number;
  start_time?: string;
  end_time?: string;
  title?: string;
  category?: string;
  schedule_effect?: string;
  end_date?: string;
  subject?: Subject;
}

export const calendarService = {
  async list(semesterId: number): Promise<CalendarEvent[]> {
    const response = await api.get(`/semesters/${semesterId}/calendar`);
    return response.data;
  },

  async save(semesterId: number, events: Omit<CalendarEvent, "id" | "semester_id">[], mode: "replace" | "merge" = "replace"): Promise<CalendarEvent[]> {
    const response = await api.post(`/semesters/${semesterId}/calendar?mode=${mode}`, events);
    return response.data;
  },
};
